opm/simulators/linalg/bda/WellContributions.cpp

```diff
@@ -54,9 +58,14 @@ WellContributions::create(const std::string& accelerator_mode, bool useWellConn)
     }
     else if(accelerator_mode.compare("rocsparse") == 0){
         if (!useWellConn) {
-            OPM_THROW(std::logic_error, "Error rocsparse requires --matrix-add-well-contributions=true");
+#if HAVE_ROCSPARSE
+            return std::make_unique<WellContributionsRocsparse>();
+#else
+        OPM_THROW(std::runtime_error, "Cannot initialize well contributions: rocsparse is not enabled");
+#endif
```

Comment:
Seems like the backend specific headers are only needed to create the unique_ptr<WellContributions...> object. Maybe move the constructions to their own function (e.g. createWellContibutionsPtr...) and put the definitions in to separate *.cpp file that include the backend specific headers. Then the headers do not need to be included here and maybe our problem is solved.

I could have missed stuff.

Reply:
I did not manage to fix the issue as you suggested. The cuda_runtime is included also from Multisegmentwells.h which is included here, which makes it more complicated, so I fixed this (temporarly) by not allowing the simultaneous use of CUDA and HIP at the same time from the cmake.